src/watcher.js

```diff
@@ -64,61 +120,81 @@ export default function createWatcher(model, config) {
-      } else {
+      try {
+        if (!this.state.handler) return;
+
+        this.state.latestPromiseId++;
+        const promiseId = this.state.latestPromiseId;
+
+        state.onPollListeners.forEach(({ listener }) =>
+          listener({
+            id: promiseId,
+            latestBlockNumber: this.state.latestBlockNumber,
+            ...(this.retry ? { retry: this.retry } : {})
+          })
+        );
+
+        const {
+          results: { blockNumber, ...data },
+          keyToArgMap
+        } = await aggregate(this.state.model, { ...this.state.config, ws: this.state.ws, id: this.state.latestPromiseId });
+
+        if (this.state.cancelPromiseId === promiseId) return;
+
+        if (typeof this.resolveFetchPromise === 'function')
+          this.resolveFetchPromise();
+
         if (
-          this.state.latestBlockNumber === null ||
-          (this.state.latestBlockNumber !== null &&
-            blockNumber > this.state.latestBlockNumber)
+          this.state.latestBlockNumber !== null &&
+          blockNumber < this.state.latestBlockNumber
         ) {
-          this.state.latestBlockNumber = blockNumber;
-          state.onNewBlockListeners.forEach(({ listener }) =>
-            listener(blockNumber)
-          );
+          // Retry if blockNumber is lower than latestBlockNumber
+          log(`Stale block returned, retrying in ${this.state.config.staleBlockRetryWait / 1000} seconds`);
+          poll.call({ state: this.state, interval: this.state.config.staleBlockRetryWait, retry: this.retry ? this.retry + 1 : 1 });
+        } else {
+          if (
+            this.state.latestBlockNumber === null ||
+            (this.state.latestBlockNumber !== null &&
+              blockNumber > this.state.latestBlockNumber)
+          ) {
+            this.state.latestBlockNumber = parseInt(blockNumber);
+            state.onNewBlockListeners.forEach(({ listener }) =>
+              listener(this.state.latestBlockNumber)
+            );
+          }
+          const events = Object.entries(data)
+            .filter(([type, value]) => isNewState(type, value, this.state.store))
+            .map(([type, value]) => ({
+              type,
+              value,
+              args: keyToArgMap[type] || []
+            }));
+          this.state.store = { ...data };
+          this.state.keyToArgMap = { ...keyToArgMap };
+          alertListeners(events);
+          poll.call({ state: this.state });
         }
-        const events = Object.entries(data)
-          .filter(([type, value]) => isNewState(type, value, this.state.store))
-          .map(([type, value]) => ({
-            type,
-            value,
-            args: keyToArgMap[type] || []
-          }));
-        this.state.store = { ...data };
-        this.state.keyToArgMap = { ...keyToArgMap };
-        alertListeners(events);
-        poll.call({ state: this.state });
+      } catch(e) {
```

Comment:
do you think it'd be sensible to add something like an `onError` listener to the `watcher` object at some point?

Reply:
yeah, that's a great idea 👌